feat: Применить паттерн проектирования "Внедрение зависимостей"

MaXM91 opened the pull request:
- Создать интерфейсы, задающие функционал хранилищ FilmStorage, UserStorage, а также создать их реализацию InMemoryFilmStorage, InMemoryUserStorage;
- Добавить поле userFriends в User для хранения друзей, добавить поле usersLike в Film для хранения лайков от пользователей;
- Добавить классы-сервисы FilmService и UserService;
- Добавить общий обработчик исключений ErrorHandler;
- Добавить класс ErrorResponse для вывода форматируемого, легкообрабатываемого сообщения об ошибке;

Commit:
feat: Применить паттерн проектирования "Внедрение зависимостей"
- Создать интерфейсы, задающие функционал хранилищ FilmStorage, UserStorage, а также создать их реализацию InMemoryFilmStorage, InMemoryUserStorage;
- Добавить поле userFriends в User для хранения друзей, добавить поле userLike в Film для хранения лайков от пользователей;
- Добавить классы-сервисы FilmService и UserService;
- Добавить общий обработчик исключений ErrorHandler;
- Добавить класс ErrorResponse для вывода форматируемого, легкообрабатываемого сообщения об ошибке;

pattyku commented on src/main/java/ru/yandex/practicum/filmorate/validators/exceptions/ErrorHandler.java:
```diff
@@ -0,0 +1,28 @@
+
+    @ExceptionHandler
+    @ResponseStatus(HttpStatus.INTERNAL_SERVER_ERROR)                          // 500
+    public ErrorResponse badWork(final Throwable exc) {
```
Throwable лучше не проверять, туда входят ошибки на уровне виртуальной машины, которые в самом приложении нереально будет исправить и обработать, поэтому лучше в качестве верхнего уровня оставить Exception
т.е. (final Throwable e) заменить на (final Exception e)

pattyku commented on src/main/java/ru/yandex/practicum/filmorate/storage/user/InMemoryUserStorage.java:
```diff
@@ -0,0 +1,77 @@
+// Work with friends list
+///////////////////////////////////////////////////////////////////////////////////////////////////
+    @Override
+    public void addFriend(Integer id1, Integer id2) {
```
Давай в наименовании не использовать цифры. Такое может потребоваться в редких случаях и задача должна подразумевать такое наименование. Например в математических задачах х1 и х2
В контролле уже есть подсказка 
`public void addFriend(Integer id, Integer friendId)`

исправить везде

pattyku commented on src/main/java/ru/yandex/practicum/filmorate/controllers/FilmController.java:
```diff
@@ -1,49 +1,60 @@
 public class FilmController {
-    private int filmId = 1;
-    private static final HashMap<Integer, Film> films = new HashMap<>();
+    FilmService filmService;
```
А где модификаторы доступа, это поле должно быть еще и обязательным
private final

исправить нужно везде

pattyku commented on src/main/java/ru/yandex/practicum/filmorate/controllers/FilmController.java:
```diff
@@ -1,49 +1,60 @@
-        films.put(film.getId(), film);
-        return film;
+    @Autowired
+    FilmController(FilmService filmService) {
```
Также можешь использовать аннотацию @requiredargsconstructor, который сделает такой же конструктор, но если появятся дополнительные зависимости, не придётся его переписывать, потому что все сгенерируется автоматически
Чтоб работало поля должны быть final

Для сведения. Выбираешь способ, который нравится и удобнее. Не все проекты используют ломбок

Commit:
style:
- изменить наименование аргументов;
- применить конструктор @RequiredArgsConstructor;
- изменить класс в обработчике под выдачу ошибки 500 с Throwable на Exception;
